Refactor <Input> to be clearer. Add namespace name validator

Nokel81 opened the pull request:
- Changes support of async validators to only run on submit
- Debounces validation to only run 500ms after the last edit.

Signed-off-by: Sebastian Malton <sebastian@malton.name>

fixes #2727 

Videos:

https://user-images.githubusercontent.com/8225332/117486490-b7fff880-af37-11eb-802c-a3f4c30d7257.mov

Note: that validator at the end is purely for the demo and is not in this PR.

Nokel81 committed:
Refactor <Input> to be clearer. Add namespace name validator

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
make asyncValidators run only before onSubmit

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jim-docker commented on src/renderer/components/input/input.tsx:
```diff
@@ -106,125 +117,105 @@ export class Input extends React.Component<InputProps, State> {
-    // run validation
-    this.validate();
-  }
+  runValidators = debounce(() => this.runValidatorsRaw(this.getValue()), 500, {
```
This is tricky. Debouncing helps when entering a url, you want to wait till the user is done typing to validate the input. But in the case of an invalid character I would prefer to be told I entered an invalid character immediately. Since you added the `asyncValidator`, which is run only before submitting (and could be used to check for a well-formed url, etc), then you don't really need the debouncing?

Nokel81 replied:
I think some form of denouncing is a good idea. That is because what is considered an "invalid character" depends on the final version.

For instance, a namespace must be a DNS label. Which cannot end in a `-`. I as a user would find it rather annoying to be told that while I am typing unless I pause. Because if I am typing slowly (but within that denounce window) I shouldn't see an error between when I type the `-` and when I type an `a` for instance.

jim-docker replied:
Still sounds to me like, for such a type, you should have 2 validators, one to check for invalid characters in real-time while the user is typing (no debouncing), and the other on-submit to check for well-formed input, such as no trailing `-`

Nokel81 replied:
I guess we could do that too. I was trying to keep it as consistent with the old design, while also acknowledging that `async` validators (currently) don't generally support aborting.

jim-docker reviewed (commented):
Doesn't seem quite right to leave the error message when the input field is returned to its initial, empty state (for which there was, correctly, no error message). Though I can see why you might leave it there.
 
https://user-images.githubusercontent.com/40840436/117495316-75441d80-af43-11eb-8b1d-de134d1fa51f.mov

Nokel81 committed:
Fix Input not working when not provided with an AsyncValidator

- Added logging by default for non-CI environments running integration
  tests

- Increasing the timeout on integration tests so that spectron is more
  likely to error out instead (should help with debugging tests)

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
don't make the logs folders on CI

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
fix bad rebase

Signed-off-by: Sebastian Malton <sebastian@malton.name>
